include/util/dynamic_graph.hpp

```diff
@@ -309,6 +332,36 @@ template <typename EdgeDataT> class DynamicGraph
         return current_iterator;
     }
 
+    void Renumber(const std::vector<NodeID> &old_to_new_node)
+    {
+        // permutate everything but the sentinel
+        util::inplacePermutation(node_array.begin(), std::prev(node_array.end()), old_to_new_node);
+
+        // Build up edge permutation
+        auto new_edge_index = 0;
+        std::vector<EdgeID> old_to_new_edge(edge_list.size(), SPECIAL_EDGEID);
+        for (auto node : util::irange<NodeID>(0, number_of_nodes))
+        {
+            auto new_first_edge = new_edge_index;
+            // move all filled edges
+            for (auto edge : GetAdjacentEdgeRange(node))
+            {
+                edge_list[edge].target = old_to_new_node[edge_list[edge].target];
+                old_to_new_edge[edge] = new_edge_index++;
+            }
+            // and all adjacent empty edges
+            for (auto edge = EndEdges(node); edge < number_of_edges && isDummy(edge); edge++)
```

Comment:
just don't understand why dummy edges are needed? If it is only to have correct results of `GetAdjacentEdgeRange` and `EndEdges` then it is better to make an iteration over `node_array` directly and remove empty edges.